### DISCUSSION TITLE
Not sure this is a bug or not...

### DISCUSSION THREAD

**Discussion by xareelee** [python/typing#1155] [General]
```python
from typing import Callable

def f(a: int, b: int) -> int:
    return a + b

g: Callable[[int, int], float] = f

d = g(1, 2)

(reveal_type(d) if typing.TYPE_CHECKING else None)

print(type(d))
```

output (with mypy):

```
# mypy
note: Revealed type is "builtins.float"

# runtime
<class 'int'>
```

The type checker told me `d` was `float`. The runtime told me `d` was `int`.

The function signatures are different with `f` and `g`, and mypy did not complain about this.

**Comment by JelleZijlstra** [comment]
There's no bug here. `int` is a subtype of `float` to the type checker (this is specified in PEP 484). The return type of a callable is covariant, so it's OK to assign a callable with a more precise type (`int`) to a less precise type (`float`). If you do that, the type checker will infer a more general type than the runtime type, as you observe.

**Reply by xareelee** [reply]
Is that possible to make type checker complaint this?

**Reply by Gobot1234** [reply]
Not really. Why would you want it to? Float can be interchanged with an int in most places
```py
In [1]: set(float.__dict__) - set(int.__dict__)
Out[1]: {'__getformat__', '__set_format__', 'fromhex', 'hex', 'is_integer'}
```

**Reply by erictraut** [reply]
Can you be more specific about would you like the type checker to complain about? As @JelleZijlstra said, this is type safe code.

Note that you'll see different behavior from mypy if you separate the type annotation and assignment of `g` on different lines:

```python
g: Callable[[int, int], float]
g = f

reveal_type(g(1, 2)) # int
```

Mypy applies type narrowing on assignment in this case, but it doesn't in the case that the assignment occurs on the same line as the type annotation (as shown in your example above). This inconsistency has been pointed out many times, but it hasn't yet been fixed in mypy. Other type checkers like pyright and pyre are consistent and always apply type narrowing on assignment.

**Reply by xareelee** [reply]
Thanks for reply. This just surprises me. 

`int` is not a subclass of `float`:

```python
print(issubclass(int, float))    # False
```

How do I check if `a` is a subtype of `b`? Is there any function to do that?
